### ISSUE TITLE
complete: keep pum visible when pressing <BS> for complete()

### ISSUE THREAD

**Issue by seblyng** [vim/vim#15833]
**Describe the solution you'd like**
First of all, I am not sure if this should be considered a bug or a feature request, so forgive me if I opened a wrong type of issue😅

When using `complete()` and showing the pum, it is removed when pressing `<BS>`. This is not the same behaviour as using omnifunc and other CTRL-X submodes. Then the pum will still be visible even after pressing `<BS>`

**Describe alternatives you've considered**
I would like it to behave the same when completing via `complete()`, and have the pum still be visible after a `<BS>`

**Additional context**
I have opened an issue over in neovim (https://github.com/neovim/neovim/issues/30723) and have added a patch there that seems to fix the problem. As this code is shared between neovim and vim, I thought I would open an issue here as well to see what you think. Is it reasonable to unify the behaviour, and make `complete()` work in the same way as the other CTRL-X submodes? If so, does the following patch seem reasonable?

```diff
diff --git a/src/insexpand.c b/src/insexpand.c
index eb920a751..24258ec72 100644
--- a/src/insexpand.c
+++ b/src/insexpand.c
@@ -1917,7 +1917,6 @@ ins_compl_bs(void)
     // Respect the 'backspace' option.
     if ((int)(p - line) - (int)compl_col < 0
 	    || ((int)(p - line) - (int)compl_col == 0 && !ctrl_x_mode_omni())
-	    || ctrl_x_mode_eval()
 	    || (!can_bs(BS_START) && (int)(p - line) - (int)compl_col
 							- compl_length < 0))
 	return K_BS;
```

Using a vimrc with the snippet from `:h complete()`

Before patch:

https://github.com/user-attachments/assets/56c4b332-6a60-4493-8315-709f4b86df57

After patch:

https://github.com/user-attachments/assets/56e88952-fc6b-4a4d-ad13-02036da72e8e

**Comment by chrisbra** [comment]
I like the consistency this would bring, so yes seems reasonable.

**Comment by seblj** [comment]
Perfect! I can submit a PR. Do I need to add a test?

**Comment by glepnir** [comment]
c_CTRL-L behavior in tests will break

**Comment by seblj** [comment]
Yeah that should be updated then. It seems like this will fix the TODO in the test as well if I read it correctly. This will now become "January" and not "Januar" with the patch

**Comment by glepnir** [comment]
yes the todo seesms like fixed then but there is about C-L behavior.  seemsl ike it not genereate new compl_match_array it's here https://github.com/vim/vim/blob/0fe17f8ffbd2588ecd2bf42dced556897bc64f89/src/edit.c#L644-L662

can give it a try to see if there are other tests that are broken.

**Comment by chrisbra** [comment]
Ah, I think that test even came from me a long time ago. Since this behaviour seems to be fixed and now also seem to work according to documentation, I think this patch makes sense. Can you create a PR for this change? @glepnir did you test it already?

**Comment by seblj** [comment]
Yes I will create a PR when I am back home!

**Comment by glepnir** [comment]
just that case which i mentionad failed. need to look at the historical PR to see the description.   comment of test case is meaningful

**Comment by puremourning** [comment]
There's a high likelihood this change of behaviour will break completion systems like YCM.

**Comment by puremourning** [comment]
Cc @bstaletic
